docs/SERVER-API.md

```diff
@@ -0,0 +1,214 @@
+This document details the API offered by the server to clients. Each method description contains:
+- The method's name
+- The method's purpose and intended effect on the server and/or the appearance of the screen rendered by zellij.
+- The method's parameters
+- The response that the client can expect to receive, including any possible error messages. N.B. Response does not detail side-effects of the request, in particular changes in what's rendered on screen.
+
+The document does not describe implementation details on the server side and is intended for use by those implementing new zellij clients.
+
+
+## Basic controls
+
+### Write input
+
+**Purpose:** Writes a set of bytes to the terminal, not to be interpeted by zellij.
+
+**Parameters:** Vector of bytes to be written.
+
+**Response:** None.
+
+### Switch input mode
+@@@Is this part of the API or a client detail?
+
+**Purpose:** Switches between different zellij [input modes](./TERMINOLOGY.md#input_modes).
+
+**Parameters:** Which input mode to switch to.
+
+**Response:** None.
+
+### Scroll up
+
+**Purpose:** Scrolls up (back) in the screen buffer of the focus [pane](./TERMINOLOGY.md#pane).
+
+**Parameters:** None.
+
+**Response:** None.
+
+### Scroll down
+
+**Purpose:** Scrolls down (forwards) in the screen buffer of the focus [pane](./TERMINOLOGY.md#pane).
+
+**Parameters:** None.
+
+**Response:** None.
+
+### Terminal resize
+
+**Purpose:** Indicates that the terminal has been resized and so the layout should be redrawn.
+
+**Parameters:** New size of the layout.
+
+**Response:** @@@do we have a lock here?
+
+### Open file
+
+**Purpose:** Indicates that a new terminal should be spawned.
+
+**Parameters:** File name (optional) - the name of a file to be opened, using a terminal application determined by first the `EDITOR` and (if `EDITOR` is not set) `VISUAL` environment variables.
+
+**Response:** @@@do we have a lock here?
+
+## Pane controls
+
+### New pane
+
+**Purpose:** Opens a new [pane](./TERMINOLOGY.md#pane).
+
+**Parameters:** Direction (optional) - the position of the new pane relative to the focus pane.
+
+**Response:** Done opening new pane - indicates that the server has finished opening the pane. This exists for the purpose of synchronization @@@what's blocked?
```

Comment:
@imsnif I don't understand the various locks we have between client and server. Is it something that the client should care about at all? Should a client block in any way? If not, then we probably want to isolate this to the server side.

Reply:
This is actually a good point. We should not try to create atomicity in operations across processes. Instead of blocking the input thread on the client-side, we should consider blocking the router thread (only if we decide to have a separate router thread for each client).

Reply:
I agree with @kunalmohan. I think blocking the router thread on the server side is a good solution, and we can probably get away with just 1 kind of lock. As long as we make sure that one action that would cause a lock doesn't spawn another action that can release it.

eg. If we want the "new tab" action to lock the router thread until the new tab is open, we shouldn't allow the "new pane" action to release the lock, because the "new pane" option can be taken as a result of the "new tab" action. In the current state of affairs this is a non-issue, but if we imagine opening a new tab with a complex 10+ pane layout, this might become one...